Source/JavaScriptCore/runtime/ProxyObject.cpp

```diff
@@ -82,6 +86,64 @@ void ProxyObject::finishCreation(VM& vm, JSGlobalObject* globalObject, JSValue t
     internalField(Field::Handler).set(vm, this, handler);
 }
 
+JSObject* ProxyObject::getHandlerTrap(JSGlobalObject* globalObject, JSObject* handler, CallData& callData, const Identifier& ident, HandlerTrap trap)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    auto ensureIsCallable = [&](JSValue value) -> JSObject* {
+        if (value.isUndefinedOrNull())
+            return nullptr;
+
+        callData = JSC::getCallData(value);
+        if (callData.type == CallData::Type::None) {
+            throwTypeError(globalObject, scope, makeString("'", String(ident.impl()), "' property of a Proxy's handler should be callable"));
+            return nullptr;
+        }
+
+        return asObject(value);
+    };
+
+    JSObject* objectPrototype = globalObject->objectPrototype();
+
+    bool isCacheValid = handler->structureID() == m_handlerStructureID.value() && objectPrototype->structureID() == m_handlerPrototypeStructureID.value();
+    if (isCacheValid) {
+        PropertyOffset offset = m_handlerTrapsOffsetsCache[static_cast<uint8_t>(trap)];
+        if (offset == invalidOffset)
+            return nullptr;
+        if (offset != emptyHandlerTrapCache)
+            return ensureIsCallable(handler->getDirect(offset));
+    } else
+        clearHandlerTrapsOffsetsCache();
+
+    PropertySlot slot(handler, PropertySlot::InternalMethodType::Get);
+    bool hasProperty = handler->getPropertySlot(globalObject, ident, slot);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    bool isSlotCacheable = slot.isUnset() || (slot.isCacheableValue() && slot.slotBase() == handler);
+    if (isSlotCacheable) {
+        bool isHandlerPrototypeChainCacheable = handler->type() == FinalObjectType && !handler->structure()->isDictionary() && handler->getPrototypeDirect() == objectPrototype && !objectPrototype->structure()->isDictionary();
```

Comment:
Need to check `JSGlobalObject::objectPrototypeChainIsSane` too, and add test changing `Object.prototype.__proto__`.

Reply:
Cacheability depends on unset case / set case.
1. If this is set case, then we can cache offset even if `handler->getPrototypeDirect() == objectPrototype` is not met because this is own property case. And cacheable dictionary works since it is own hitting property. We should check uncacheable dictionary in this case instead.
2. If this is unset case, then we need to check dictionary condition, and prototype chains.

Reply:
But since we would like to reflect all three fields with this one cache, the above different would not be necessary. But anyway, we need to check `objectPrototypeChainIsSane`

Reply:
Discussed `objectPrototypeChainIsSane` thing offline with Yusuke: it's not actually necessary since `Object.prototype` is an immutable prototype object.

Reply:
As for cacheability condition, we can certainly make it more sophisticated, but even the current one covers 99.9% cases so why bother.